### PR TITLE
Command pattern

### PR DESCRIPTION
**Pull request by xredo**
This code shows the command pattern the way we defined it in the last Martian Day.

### CODE REVIEW

File: rails/commands-pattern/app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+class UsersController < ApplicationController
+
+  def create
+    authorize User
+
+    Users::Create.new.call(user_params).then do |result|
```
Comment (Dor3nz) [inline rails/commands-pattern/app/controllers/users_controller.rb]: `Users::Create.new(user_params).call` ?

Comment (xredo) [reply rails/commands-pattern/app/controllers/users_controller.rb]: 😲

Nice catch 👍

**Review by ItsGenis — APPROVED** [review]
As  I already discussed with @oriolbcn during the retreat, I like the approach. :+1:

**Review by artero — APPROVED** [review]

File: rails/commands-pattern/app/commands/users/create.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# Notice that commands are scoped into a module referencing the entity they are working
+# with. There are situations where the command might not refer to a single entity.
+# The name of the module doesn't need to match the name of a Rails model.
+module Users
+  class Create
```
Comment (artero) [inline rails/commands-pattern/app/commands/users/create.rb]: I think we should use a suffix in the class name like `CreateCommand`, In Rails all classes except models have a suffix and other gems like Keynote uses prefixes in its classes too. Use suffixes could be useful for the project's newbies too.

We could think we only should use suffix if the class have an inheritance but in the Form Objects patter we use Object without inheritance and its classes have suffix too.

**Review by dvdgs88 — APPROVED** [review]

**Review by davidgg — APPROVED** [review]

File: rails/commands-pattern/app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+class UsersController < ApplicationController
+
+  def create
+    authorize User
+
+    Users::Create.new(user_params).call.then do |result|
```
Comment (danigb) [inline rails/commands-pattern/app/controllers/users_controller.rb]: I think `Create` is not a good example (or at least complete). When using `Update` you have to pass two arguments, and then there are two options:

```ruby
Users::Update.new(user).call(params)
```

or 

```ruby
Users::Update.new(user, params).call
```

I don't know which one to use 🤷‍♀️

**Review by danigb — CHANGES_REQUESTED** [review]
Finally I started to stop worring and liking this pattern 🙄 ... I have doubts about how to pass params to the command (constructor or `call` method or both). I leaved a comment about it

Also I'm not convinced about `then` approach.

Finally, it has to be converted to markdown style pattern 😇

File: rails/commands-pattern/app/controllers/users_controller.rb (hunk 3 above)
Comment (danigb) [inline rails/commands-pattern/app/controllers/users_controller.rb]: I'm not convinced about using `then`. There are two problems:
1. Node developers associate `then` to some kind of asynchronous execution. This is a ruby problem.
2. The code is IMO unnecessarily complex. I prefer to read:

```ruby
result = Users::Create.new(user_params).call
if result.ok
...
```

**Pull request merged**